### PR TITLE
Custom distribution sampling

### PR DESCRIPTION
**Pull request by GrimFe**

### CODE REVIEW

File: sandy/core/cov.py [hunk 1]
```diff
@@ -561,21 +569,64 @@ def sampling(self, nsmp, seed=None, rows=None, pdf='normal',
...
+        1   3.00053e+00
+        dtype: float64
         """
         dim = self.data.shape[0]
         y = sample_distribution(dim, nsmp, seed=seed, pdf=pdf)
         y = sps.csc_matrix(y)
         L = sps.csr_matrix(self.get_L(rows=rows, tolerance=tolerance))
-        samples = L.dot(y).toarray()
+        if relative:
+            samples = (L.dot(y).toarray() + 1) * mean
```
Comment (AitorBengoechea) [inline sandy/core/cov.py:627]: Samples = (samples+1)*mean

File: sandy/core/cov.py [hunk 2]
```diff
@@ -561,21 +569,64 @@ def sampling(self, nsmp, seed=None, rows=None, pdf='normal',
...
+        else:
+            samples = L.dot(y).toarray() + mean
```
Comment (AitorBengoechea) [inline sandy/core/cov.py:629]: same as above

File: sandy/core/cov.py [hunk 3]
```diff
@@ -2301,21 +2352,28 @@ def triu_matrix(matrix, kind='upper'):
     return CategoryCov(pd.DataFrame(values, index=index, columns=columns))
 
 
-def sample_distribution(dim, nsmp, seed=None, pdf='normal'):
+def sample_distribution(dim, nsmp, mean=0, seed=None, pdf='normal'):
```
Comment (AitorBengoechea) [inline sandy/core/cov.py:2355]: This mean and the mean in the method `sampling` are always different or they had to be same?

File: sandy/core/cov.py [hunk 4]
```diff
@@ -509,29 +509,37 @@ def invert(self, rows=None):
         M_inv = M_inv.reindex(index=index, columns=columns).fillna(0)
         return self.__class__(M_inv)
 
-    def sampling(self, nsmp, seed=None, rows=None, pdf='normal',
-                 tolerance=None, threshold=None):
+    def sampling(self, nsmp, seed=None, rows=None, mean=1, pdf='normal',
+                 tolerance=None, threshold=None, relative=True):
```
Comment (luca-fiorito-11) [inline sandy/core/cov.py:513]: As mentioned in another PR. I am not sure how useful this `threshold` card can be

File: sandy/core/cov.py [hunk 5]
```diff
@@ -561,21 +569,64 @@ def sampling(self, nsmp, seed=None, rows=None, pdf='normal',
...
+        0   2.99968e+00
+        1   3.00053e+00
+        dtype: float64
         """
         dim = self.data.shape[0]
         y = sample_distribution(dim, nsmp, seed=seed, pdf=pdf)
         y = sps.csc_matrix(y)
         L = sps.csr_matrix(self.get_L(rows=rows, tolerance=tolerance))
-        samples = L.dot(y).toarray()
+        if relative:
```
Comment (luca-fiorito-11) [inline sandy/core/cov.py:626]: if we have relative and absolute, do we need to also have `mean`?

File: sandy/core/cov.py (hunk 1 above)
Comment (luca-fiorito-11) [inline sandy/core/cov.py:627]: I don't like running twice `L.dot(y).toarray()`

File: sandy/core/cov.py [hunk 6]
```diff
@@ -540,18 +548,18 @@ def sampling(self, nsmp, seed=None, rows=None, pdf='normal',
...
         Examples
         --------
-        Draw 3 sets of samples using custom seed.
+        Draw 3 sets of samples using custom seed:
         >>> sandy.CategoryCov([[1, 0.4],[0.4, 1]]).sampling(3, seed=11)
                      0            1
-        0 -1.74945e+00 -3.13159e+00
-        1  2.86073e-01  1.06836e-01
-        2  4.84565e-01 -9.91209e-02
+        0 -7.49455e-01 -2.13159e+00
```
Comment (luca-fiorito-11) [inline sandy/core/cov.py:554]: why are the samples different? it doesn't seem to me there is any substantial change to this algorithm

File: sandy/core/cov.py (hunk 3 above)
Comment (luca-fiorito-11) [reply sandy/core/cov.py:2355]: I am not sure we should even have `mean` here.

File: sandy/core/cov.py [hunk 7]
```diff
@@ -2327,13 +2385,42 @@ def sample_distribution(dim, nsmp, seed=None, pdf='normal'):
...
+    3.00025
+
+    >>> sandy.cov.sample_distribution(2, 1000000, seed=11, mean=3).std().round(5)
+    0.99919
     """
     np.random.seed(seed=seed)
     if pdf == 'normal':
         y = np.random.randn(dim, nsmp)
-    return y
+    elif pdf == 'lognormal':
```
Comment (luca-fiorito-11) [inline sandy/core/cov.py:2415]: can we get a visual test on a separate notebook submitted to https://luca-fiorito-11.github.io/sandy_notebooks/

**Review by luca-fiorito-11 — CHANGES_REQUESTED** [review]

File: sandy/core/cov.py (hunk 4 above)
Comment (luca-fiorito-11) [reply sandy/core/cov.py:513]: It will be kept in mind. But we will remove it!

Comment (luca-fiorito-11) [reply sandy/core/cov.py:513]: It will be kept in mind. But we will remove it!

File: sandy/core/cov.py [hunk 8]
```diff
@@ -509,29 +509,36 @@ def invert(self, rows=None):
...
         return self.__class__(M_inv)
 
-    def sampling(self, nsmp, seed=None, rows=None, pdf='normal',
-                 tolerance=None, threshold=None):
+    def get_perturbation_coeff(self, nsmp, seed=None, rows=None, pdf='normal',
+                               tolerance=None, relative=True):
         """
-        Extract random samples from normal distribution centered in zero
-        and with given covariance matrix.
+        Extract perturbation coefficients according to chosed distribution with
```
Comment (luca-fiorito-11) [inline sandy/core/cov.py:515]: chosen

File: sandy/core/cov.py [hunk 9]
```diff
@@ -509,29 +509,36 @@ def invert(self, rows=None):
...
+        covariance from given covariance matrix (self).
```
Comment (luca-fiorito-11) [inline sandy/core/cov.py:516]: Remove (self)

File: sandy/core/cov.py [hunk 10]
```diff
@@ -509,29 +509,36 @@ def invert(self, rows=None):
         M_inv = M_inv.reindex(index=index, columns=columns).fillna(0)
         return self.__class__(M_inv)
 
-    def sampling(self, nsmp, seed=None, rows=None, pdf='normal',
-                 tolerance=None, threshold=None):
+    def get_perturbation_coeff(self, nsmp, seed=None, rows=None, pdf='normal',
```
Comment (luca-fiorito-11) [inline sandy/core/cov.py:512]: We will call it sampling or get_samples. This is consistent with standard nomenclature. 
Perturbations will get mixed up with `custom_perturbation`.
These are anyway samples!

File: sandy/core/cov.py [hunk 11]
```diff
@@ -2327,12 +2363,35 @@ def sample_distribution(dim, nsmp, seed=None, pdf='normal'):
     >>> sandy.cov.sample_distribution(2, 3, seed=11)
     array([[ 1.74945474, -0.286073  , -0.48456513],
            [-2.65331856, -0.00828463, -0.31963136]])
+
+    >>> sandy.cov.sample_distribution(2, 3, seed=11, pdf='lognormal')
```
Comment (luca-fiorito-11) [inline sandy/core/cov.py:2367]: sample_distribution should be added to `__all__` and be called as `sandy.sample_distribution`

File: sandy/core/cov.py [hunk 12]
```diff
@@ -2327,12 +2363,35 @@ def sample_distribution(dim, nsmp, seed=None, pdf='normal'):
...
+    array([[-1.10757829, -1.66458659, -0.12741476],
+           [ 0.77919399, -0.27642282, -0.05048201]])
+
     >>> sandy.cov.sample_distribution(2, 1000000, seed=11).mean().round(5)
     0.00025
+
+    >>> sandy.cov.sample_distribution(2, 1000000, seed=11, pdf='lognormal').mean().round(5)
+    -0.00047
+
+    >>> sandy.cov.sample_distribution(2, 1000000, seed=11, pdf='uniform').mean().round(5)
```
Comment (luca-fiorito-11) [inline sandy/core/cov.py:2381]: What about standard deviations and correlations?

File: sandy/core/cov.py [hunk 13]
```diff
@@ -2327,12 +2363,35 @@ def sample_distribution(dim, nsmp, seed=None, pdf='normal'):
...
+
+    >>> sandy.cov.sample_distribution(2, 3, seed=11, pdf='uniform')
+    array([[-1.10757829, -1.66458659, -0.12741476],
+           [ 0.77919399, -0.27642282, -0.05048201]])
+
     >>> sandy.cov.sample_distribution(2, 1000000, seed=11).mean().round(5)
     0.00025
+
+    >>> sandy.cov.sample_distribution(2, 1000000, seed=11, pdf='lognormal').mean().round(5)
+    -0.00047
```
Comment (luca-fiorito-11) [inline sandy/core/cov.py:2379]: What about standard deviations and correlations?

File: sandy/core/cov.py [hunk 14]
```diff
@@ -2327,12 +2363,35 @@ def sample_distribution(dim, nsmp, seed=None, pdf='normal'):
...
+    -0.00115
     """
     np.random.seed(seed=seed)
     if pdf == 'normal':
         y = np.random.randn(dim, nsmp)
+    elif pdf == 'lognormal':
+        sl = ml = 1  # target std and mean of lognormal distribution
+        mn = 2 * np.log(ml) - .5 * np.log(sl**2 + np.exp(2 * np.log(ml)))
+        sn = np.sqrt(2 * (np.log(ml) - mn))
+        y = np.random.lognormal(mn, sn, (dim, nsmp)) - ml
```
Comment (luca-fiorito-11) [inline sandy/core/cov.py:2391]: I don't like transposing it to 0. Then it is not lognormal anymore

**Review by luca-fiorito-11 — CHANGES_REQUESTED** [review]

**Review by luca-fiorito-11 — CHANGES_REQUESTED** [review]

**Pull request merged**